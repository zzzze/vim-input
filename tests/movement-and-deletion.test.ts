import { describe, expect, it } from 'vitest'
import { ElementController } from './utils/element-controller'
import { prepare } from './utils/prepare'
import delay from './utils/delay'

describe('movement and deletion (dd)', () => {
  it('should delete first line', async () => {
    const { dom, element } = await prepare('tests/fixtures/letter.html')
    const controller = new ElementController(dom, element, dom.window.app.vim)
    controller.focus()
    await controller.pressKey('Escape')
    await controller.pressKey('d', { repeatCount: 2 })
    expect(element.value).toMatchSnapshot()
    expect(dom.window.app.vim.currentMode).to.equal('general_mode')
  })

  it('should delete second line', async () => {
    const { dom, element } = await prepare('tests/fixtures/letter.html')
    const controller = new ElementController(dom, element, dom.window.app.vim)
    controller.focus()
    await controller.pressKey('Escape')
    await controller.pressKey('j')
    await controller.pressKey('d', { repeatCount: 2 })
    expect(element.value).toMatchSnapshot()
    expect(dom.window.app.vim.currentMode).to.equal('general_mode')
  })

  it('should delete last line', async () => {
    const { dom, element } = await prepare('tests/fixtures/letter.html')
    const controller = new ElementController(dom, element, dom.window.app.vim)
    controller.focus()
    await controller.pressKey('Escape')
    await controller.pressKey('shift+g')
    await controller.pressKey('d', { repeatCount: 2 })
    expect(element.value).toMatchSnapshot()
    expect(dom.window.app.vim.currentMode).to.equal('general_mode')
  })
})

describe('movement and deletion (x)', () => {
  it('should delete first character of first line', async () => {
    const { dom, element } = await prepare('tests/fixtures/letter.html')
    const controller = new ElementController(dom, element, dom.window.app.vim)
    controller.focus()
    await controller.pressKey('Escape')
    await controller.pressKey('x')
    expect(element.value).toMatchSnapshot()
    expect(dom.window.app.vim.currentMode).to.equal('general_mode')
  })

  it('should not delete a empty line', async () => {
    const { dom, element } = await prepare('tests/fixtures/letter.html')
    const controller = new ElementController(dom, element, dom.window.app.vim)
    controller.focus()
    await controller.pressKey('Escape')
    await controller.pressKey('j')
    await controller.pressKey('x', { repeatCount: 10 })
    expect(element.value).toMatchSnapshot()
    expect(dom.window.app.vim.currentMode).to.equal('general_mode')
  })

  it('should delete second character', async () => {
    const { dom, element } = await prepare('tests/fixtures/letter.html')
    const controller = new ElementController(dom, element, dom.window.app.vim)
    controller.focus()
    await controller.pressKey('Escape')
    await controller.pressKey('l')
    await controller.pressKey('x')
    expect(element.value).toMatchSnapshot()
    expect(dom.window.app.vim.currentMode).to.equal('general_mode')
  })
})

describe('movement and deletion (s)', () => {
  it('should delete first character of first line', async () => {
    const { dom, element } = await prepare('tests/fixtures/letter.html')
    const controller = new ElementController(dom, element, dom.window.app.vim)
    controller.focus()
    await controller.pressKey('Escape')
    await controller.pressKey('s')
    expect(element.value).toMatchSnapshot()
    await delay(200)
    expect(dom.window.app.vim.currentMode).to.equal('edit_mode')
  })

  it('should not delete a empty line', async () => {
    const { dom, element } = await prepare('tests/fixtures/letter.html')
    const controller = new ElementController(dom, element, dom.window.app.vim)
    controller.focus()
    await controller.pressKey('Escape')
    await controller.pressKey('j')
    await controller.pressKey('s')
    expect(element.value).toMatchSnapshot()
    await delay(200)
    expect(dom.window.app.vim.currentMode).to.equal('edit_mode')
  })
})
