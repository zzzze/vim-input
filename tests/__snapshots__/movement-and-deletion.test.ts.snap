// Vitest Snapshot v1

exports[`movement and deletion (dd) > should delete first line 1`] = `
"
I hope this letter finds you well. It's been a while since we've spoken, and I wanted to take this opportunity to catch up with you.

First of all, I wanted to congratulate you on your recent promotion! It's well-deserved, and I know you've worked hard to get to where you are. I'm really happy for you.

Secondly, I wanted to tell you about some exciting news on my end. I've decided to start my own business! It's been a dream of mine for a long time, and I finally took the leap. It's been a lot of work, but also very rewarding.

Anyway, I'd love to hear what's new with you. Let's catch up soon over a cup of coffee?

Take care,

Jane
"
`;

exports[`movement and deletion (dd) > should delete last line 1`] = `
"Dear John,

I hope this letter finds you well. It's been a while since we've spoken, and I wanted to take this opportunity to catch up with you.

First of all, I wanted to congratulate you on your recent promotion! It's well-deserved, and I know you've worked hard to get to where you are. I'm really happy for you.

Secondly, I wanted to tell you about some exciting news on my end. I've decided to start my own business! It's been a dream of mine for a long time, and I finally took the leap. It's been a lot of work, but also very rewarding.

Anyway, I'd love to hear what's new with you. Let's catch up soon over a cup of coffee?

Take care,

"
`;

exports[`movement and deletion (dd) > should delete second line 1`] = `
"Dear John,
I hope this letter finds you well. It's been a while since we've spoken, and I wanted to take this opportunity to catch up with you.

First of all, I wanted to congratulate you on your recent promotion! It's well-deserved, and I know you've worked hard to get to where you are. I'm really happy for you.

Secondly, I wanted to tell you about some exciting news on my end. I've decided to start my own business! It's been a dream of mine for a long time, and I finally took the leap. It's been a lot of work, but also very rewarding.

Anyway, I'd love to hear what's new with you. Let's catch up soon over a cup of coffee?

Take care,

Jane
"
`;

exports[`movement and deletion (s) > should delete first character of first line 1`] = `
"ear John,

I hope this letter finds you well. It's been a while since we've spoken, and I wanted to take this opportunity to catch up with you.

First of all, I wanted to congratulate you on your recent promotion! It's well-deserved, and I know you've worked hard to get to where you are. I'm really happy for you.

Secondly, I wanted to tell you about some exciting news on my end. I've decided to start my own business! It's been a dream of mine for a long time, and I finally took the leap. It's been a lot of work, but also very rewarding.

Anyway, I'd love to hear what's new with you. Let's catch up soon over a cup of coffee?

Take care,

Jane
"
`;

exports[`movement and deletion (s) > should not delete a empty line 1`] = `
"Dear John,
 
I hope this letter finds you well. It's been a while since we've spoken, and I wanted to take this opportunity to catch up with you.

First of all, I wanted to congratulate you on your recent promotion! It's well-deserved, and I know you've worked hard to get to where you are. I'm really happy for you.

Secondly, I wanted to tell you about some exciting news on my end. I've decided to start my own business! It's been a dream of mine for a long time, and I finally took the leap. It's been a lot of work, but also very rewarding.

Anyway, I'd love to hear what's new with you. Let's catch up soon over a cup of coffee?

Take care,

Jane
"
`;

exports[`movement and deletion (x) > should delete first character of first line 1`] = `
"ear John,

I hope this letter finds you well. It's been a while since we've spoken, and I wanted to take this opportunity to catch up with you.

First of all, I wanted to congratulate you on your recent promotion! It's well-deserved, and I know you've worked hard to get to where you are. I'm really happy for you.

Secondly, I wanted to tell you about some exciting news on my end. I've decided to start my own business! It's been a dream of mine for a long time, and I finally took the leap. It's been a lot of work, but also very rewarding.

Anyway, I'd love to hear what's new with you. Let's catch up soon over a cup of coffee?

Take care,

Jane
"
`;

exports[`movement and deletion (x) > should delete second character 1`] = `
"Dar John,

I hope this letter finds you well. It's been a while since we've spoken, and I wanted to take this opportunity to catch up with you.

First of all, I wanted to congratulate you on your recent promotion! It's well-deserved, and I know you've worked hard to get to where you are. I'm really happy for you.

Secondly, I wanted to tell you about some exciting news on my end. I've decided to start my own business! It's been a dream of mine for a long time, and I finally took the leap. It's been a lot of work, but also very rewarding.

Anyway, I'd love to hear what's new with you. Let's catch up soon over a cup of coffee?

Take care,

Jane
"
`;

exports[`movement and deletion (x) > should not delete a empty line 1`] = `
"Dear John,
 
I hope this letter finds you well. It's been a while since we've spoken, and I wanted to take this opportunity to catch up with you.

First of all, I wanted to congratulate you on your recent promotion! It's well-deserved, and I know you've worked hard to get to where you are. I'm really happy for you.

Secondly, I wanted to tell you about some exciting news on my end. I've decided to start my own business! It's been a dream of mine for a long time, and I finally took the leap. It's been a lot of work, but also very rewarding.

Anyway, I'd love to hear what's new with you. Let's catch up soon over a cup of coffee?

Take care,

Jane
"
`;
